divi/src/net.cpp

```diff
@@ -1453,23 +1495,43 @@ void ThreadMessageHandler()
             }
             boost::this_thread::interruption_point();
 
-            // Send messages
+            // Handle potential ping messages first.
+            if (pnode->CanSendMessagesToPeer())
             {
                 TRY_LOCK(pnode->cs_vSend, lockSend);
                 if (lockSend)
-                    g_signals.SendMessages(pnode, pnode == pnodeTrickle || pnode->fWhitelisted);
+                    pnode->MaybeSendPing();
             }
             boost::this_thread::interruption_point();
 
+            // Rebroadcasting is done for all nodes after the ping
+            // of the first; this is to mimic previous behaviour.
             if (rebroadcast) {
-                // Periodically clear setAddrKnown to allow refresh broadcasts
-                if (nLastRebroadcast)
-                    pnode->setAddrKnown.clear();
+                for (auto* pnode2 : vNodesCopy) {
+                    TRY_LOCK(pnode2->cs_vSend, lockSend);
+                    if (!lockSend)
+                        continue;
+
+                    // Periodically clear setAddrKnown to allow refresh broadcasts
+                    pnode2->setAddrKnown.clear();
```

Comment:
This line, according to the previous behaviour, should only be invoked if you've rebroadcast before. So we should check nLastRebroadcast> 0, no?

Reply:
Ah yeah indeed, I got confused with that code and thought this check was more or less redundant with the rebroadcast flag itself (which it isn't).  I'll add it back and then squash up the commits.